Sources/MacOSPlatform/MacOS.swift

```diff
@@ -0,0 +1,184 @@
+import Foundation
+import SwiftlyCore
+
+struct SwiftPkgInfo: Codable {
+    var CFBundleIdentifier: String
+}
+
+/// `Platform` implementation for macOS systems.
+public struct MacOS: Platform {
+    public init() {}
+
+    public var appDataDirectory: URL {
+        return FileManager.default.homeDirectoryForCurrentUser
+            .appendingPathComponent("Library/Application Support", isDirectory: true)
+    }
+
+    public var toolchainFileExtension: String {
+        "pkg"
+    }
+
+    public func isSystemDependencyPresent(_: SystemDependency) -> Bool {
+        // All system dependencies on macOS should be present
+        true
+    }
+
+    public func verifySystemPrerequisitesForInstall(requireSignatureValidation: Bool) throws {
+        // All system prerequisites should be there for macOS
+    }
+
+    public func install(from tmpFile: URL, version: ToolchainVersion) throws {
+        guard tmpFile.fileExists() else {
+            throw Error(message: "\(tmpFile) doesn't exist")
+        }
+
+        if !self.swiftlyToolchainsDir.fileExists() {
+            try FileManager.default.createDirectory(at: self.swiftlyToolchainsDir, withIntermediateDirectories: false)
+        }
+
+        SwiftlyCore.print("Installing package in user home directory...")
+        try runProgram("installer", "-pkg", tmpFile.path, "-target", "CurrentUserHomeDirectory")
+    }
+
+    public func uninstall(_ toolchain: ToolchainVersion) throws {
+        SwiftlyCore.print("Uninstalling package in user home directory...")
+
+        let toolchainDir = self.swiftlyToolchainsDir.appendingPathComponent("\(toolchain.identifier).xctoolchain", isDirectory: true)
+
+        let decoder = PropertyListDecoder()
+        let infoPlist = toolchainDir.appendingPathComponent("Info.plist")
+        guard let data = try? Data(contentsOf: infoPlist) else {
+            throw Error(message: "could not open \(infoPlist)")
+        }
+
+        guard let pkgInfo = try? decoder.decode(SwiftPkgInfo.self, from: data) else {
+            throw Error(message: "could not decode plist at \(infoPlist)")
+        }
+
+        try FileManager.default.removeItem(at: toolchainDir)
+
+        let homedir = ProcessInfo.processInfo.environment["HOME"]!
+        try runProgram("pkgutil", "--volume", homedir, "--forget", pkgInfo.CFBundleIdentifier)
+    }
+
+    public func use(_ toolchain: ToolchainVersion, currentToolchain: ToolchainVersion?) throws -> Bool {
```

Comment:
There's enough difference that I'm not sure what is a good abstraction here. Also, I expect that when we go to a proxy model these basically go away because there's no symlinks to retarget.

Reply:
The only difference appears to be the toolchain directory and the warning about rehashing:
```diff
--- /home/patrick/swiftly/Sources/LinuxPlatform/current.txt	2024-06-17 22:51:14.487659136 -0400
+++ /home/patrick/swiftly/macos.txt	2024-06-17 23:01:42.160478308 -0400
@@ -1,7 +1,6 @@
-
     public func use(_ toolchain: ToolchainVersion, currentToolchain: ToolchainVersion?) throws -> Bool {
         let toolchainBinURL = self.swiftlyToolchainsDir
-            .appendingPathComponent(toolchain.name, isDirectory: true)
+            .appendingPathComponent(toolchain.identifier + ".xctoolchain", isDirectory: true)
             .appendingPathComponent("usr", isDirectory: true)
             .appendingPathComponent("bin", isDirectory: true)
 
@@ -42,12 +41,22 @@
             )
         }
 
+        SwiftlyCore.print("""
+            NOTE: On macOS it is possible that the shell will pick up the system Swift on the path
+            instead of the one that swiftly has installed for you. You can run the 'hash -r'
+            command to update the shell with the latest PATHs.
+
+                hash -r
+
+            """
+        )
+
         return true
     }
 
     public func unUse(currentToolchain: ToolchainVersion) throws {
         let currentToolchainBinURL = self.swiftlyToolchainsDir
-            .appendingPathComponent(currentToolchain.name, isDirectory: true)
+            .appendingPathComponent(currentToolchain.identifier + ".xctoolchain", isDirectory: true)
             .appendingPathComponent("usr", isDirectory: true)
             .appendingPathComponent("bin", isDirectory: true)
 
@@ -57,7 +66,7 @@
             }
 
             let url = self.swiftlyBinDir.appendingPathComponent(existingExecutable)
-            let vals = try url.resourceValues(forKeys: [.isSymbolicLinkKey])
+            let vals = try url.resourceValues(forKeys: [URLResourceKey.isSymbolicLinkKey])
 
             guard let islink = vals.isSymbolicLink, islink else {
                 throw Error(message: "Found executable not managed by swiftly in SWIFTLY_BIN_DIR: \(url.path)")


```
For the toolchain directory, we can add something like the following to the `Platform` protocol:
```swift
func getToolchainDirectoryName(for toolchain: ToolchainVersion) -> String
```
On Linux, this could return `toolchain.name`, and on macOS it can return `toolchain.identifier + ".xctoolchain"`.

Reply:
I think we can leave this to a later PR, if it is needed.